File: PythonSolutions/PostOrderTraversal.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root == None:
            return []
        elif root.left == None and root.right == None:
            return [root.val]
        elif root.left == None:
            l = []
            l.extend(self.postorderTraversal(root.right))
            l.append(root.val)
            return l
        elif root.right == None:
            l = []
            l.extend(self.postorderTraversal(root.left))
            l.append(root.val)
            return l
        else:
            l = []
            l.extend(self.postorderTraversal(root.left))
            l.extend(self.postorderTraversal(root.right))
            l.append(root.val)
            return l

File: JavaSolutions/PostOrderTraversal.java
import java.util.ArrayList;

//Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class PostOrderTraversal {
    public List<Integer> postorderTraversal(TreeNode root) {
        List return_val = new ArrayList<Integer>();
        if (root == null) {
            return return_val;
        }
        else if (this.isEmpty(root)) {
            return_val.add(root.val);
        }
        else if (root.left == null) {
            return_val.addAll(postorderTraversal(root.right));
            return_val.add(root.val);
        }
        else if (root.right == null) {
            return_val.addAll(postorderTraversal(root.left));
            return_val.add(root.val);
        }
        else {
            return_val.addAll(postorderTraversal(root.left));
            return_val.addAll(postorderTraversal(root.right));
            return_val.add(root.val);
        }
        return return_val;
    }

    private boolean isEmpty(TreeNode root) {
        return root.left == null && root.right == null;
    }
}